### PR TITLE
Updating performance writer

### PR DESCRIPTION
**Pull request by adamdbrw**
This is still the "clunky" script. While the refactoring of performance writer is scheduled, this PR contains several important updates that make the performance benchmarking packages usable with recent changes: 
- uses `--storage-config-file `option with two examples (optimized and non-optimized). This enables testing of other pragma settings easily at will.
- message lost statistic is now read from metadata yaml file. This is more correct and also makes performance benchmarking ready for cache double-buffers PR 546.
- removed performance-affecting outputs (writing '.' and 'X' on writes and misses affected performance, especially with multiple instances & lots of small messages). 
- supports bag splitting in benchmarking by exposing max_bagfile_size parameter to the script.

This PR does not replace the necessary work to be done in performance packages (replacing the script with launch files, expanding parametrization), but still serves as a strict upgrade over the current state. Some of these changes should have been a part of earlier PRs.

### CODE REVIEW

**Comment by adamdbrw** [comment]
@mjeronimo this PR has some updates to benchmarking, could you take a look?

File: rosbag2_performance/rosbag2_performance_writer_benchmarking/src/writer_benchmark.cpp
Comment (mjeronimo) [inline rosbag2_performance/rosbag2_performance_writer_benchmarking/src/writer_benchmark.cpp:153]: Grammar: "Percentage of all messages that **were** successfully recorded: "

**Review by mjeronimo — COMMENTED** [review]
Looks good to me. One minor grammatical mistake/typo.

**Review by mjeronimo — APPROVED** [review]

**Comment by adamdbrw** [comment]
@mjeronimo shall we run CI and merge this if all turns out OK?

**Comment by mjeronimo** [comment]
Yes. Launched the CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13180/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8111/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10896/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13194/)

**Comment by mjeronimo** [comment]
Restart the MacOS build: [Build Status](https://ci.ros2.org/job/ci_osx/10907/)

**Comment by pijaro** [comment]
I've changed how yaml is included in `writer_benchmark.cpp` for Windows. We can rerun windows CI now and see if everything is compiling fine.

**Comment by mjeronimo** [comment]
Restarted the Windows build: [Build Status](https://ci.ros2.org/job/ci_windows/13210/)

**Comment by mjeronimo** [comment]
Background information on the pragmas added:

- https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-1-c4251?view=msvc-160
- https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-2-c4275?redirectedfrom=MSDN&view=msvc-160

The class YAML::Node is exported in a DLL, but includes std C++ objects in its interface, such as:
`  const std::string& Scalar() const;`

**Comment by mjeronimo** [comment]
Re-ran the CI since there was a code change:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13199/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8130/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10915/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13216/)

The Windows build has the expected unrelated warnings.

**Comment by mjeronimo** [comment]
@emersonknapp  Any comments before merging?

**Comment by adamdbrw** [comment]
@emersonknapp should we proceed to merge? (after resolving the conflict)

**Review by emersonknapp — APPROVED** [review]
Yes, this LGTM after conflict resolution and green CI (the windows warnings are fixed now)

**Comment by adamdbrw** [comment]
The conflicting commit, [Compress bag files in separate threads](https://github.com/ros2/rosbag2/pull/506), actually introduced some errors in rosbag2_performance_writer_benchmarking:

1. The benchmarking script was broken for empty (default) value of compression_format
1. The writer_ variable was incorrectly overwritten, see [here](https://github.com/ros2/rosbag2/pull/506/files#r542343733)

I have corrected these errors, however this also means that compression was likely not tested with the performance packages. @emersonknapp, @pjreed.

This fact is not blocking this PR AFAIK.

**Comment by pjreed** [comment]
Good catch.  I think that issue with the variable being overwritten was probably introduced as a result of merging to fix conflicts, as I'm pretty sure when I was doing performance testing that I observed compression running.  Anyway, this change looks good to me.

**Comment by mjeronimo** [comment]
Re-runing CI after latest change
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13263/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8193/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10978/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13292/)

**Comment by adamdbrw** [comment]
The warning on MacOS is concerned with an unused private member that was added in #506 - and should have been causing warnings before - I don't know why this happens only now, perhaps a compiler update or some settings changed? Anyway, an easy fix (see this [line](https://github.com/ros2/rosbag2/pull/506/files#r543301353)).

The windows fail is the usual Yaml stuff - should have known by now. Please note that the benchmarking package is not yet that usable on windows - the main utility script is not portable.

**Comment by mjeronimo** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13275/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8207/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10997/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13307/)

Restart the macOS build
[Build Status](https://ci.ros2.org/job/ci_osx/11002/)

**Pull request merged**
